Action View Form Helpers> 2. Dealing with Model Objects (#346)
campuscode/rails-guides-pt-BR: pt-BR/form_helpers.md

TIP: Se você estiver usando campos de entradas de senha (para qualquer finalidade), convêm configurar sua aplicação para impedir que esses parâmetros sejam registrados. Você pode aprender sobre este assunto no guia [Protegendo aplicações Rails](security.html#logging)

Trabalhando com Objetos *Model*
--------------------------

### *Helpers* de Objetos *Model*

> Uma tarefa particularmente comum para um formulário é editar ou criar um objeto *model*. Enquanto os *helpers* `*_tag` possam ser usados para essa tarefa, eles são um pouco verbosos, pois para cada tag você teria que garantir que o nome do parâmetro correto seja usado e atribuir o valor padrão de entrada de maneira apropriada. O Rails fornece *helpers* personalizados para essa função. Esses *helpers* não possuem o sufixo `_tag`, por exemplo `text_field`, `text_area`.

Reviewer: ```suggestion
Uma tarefa particularmente comum para um formulário é editar ou criar um objeto *model*. Embora os *helpers* `*_tag` possam ser usados para essa tarefa, eles são um pouco verbosos, pois para cada *tag* você teria que garantir que o nome do parâmetro correto seja usado e atribuir o valor padrão de entrada de maneira apropriada. O Rails fornece *helpers* personalizados para essa função. Esses *helpers* não possuem o sufixo `_tag`, por exemplo `text_field`, `text_area`.
```

> Para esses *helpers* o primeiro argumento é o nome de uma variável de instância e o segundo é o nome de um método (geralmente um atributo) para chamar esse objeto. O Rails vai definir o valor do controle de entrada para o valor de retorno daquele método para o objeto e definirá um nome de entrada apropriado. Se o seu controlador definiu `@person` e o nome dessa pessoa é Henry, um formulário contendo:

Reviewer: ```suggestion
Para esses *helpers* o primeiro argumento é o nome de uma variável de instância e o segundo é o nome de um método (geralmente um atributo) para chamar esse objeto. O Rails vai definir o valor do controle de entrada para o valor de retorno daquele método para o objeto e definirá um nome de entrada apropriado. Se o seu *controller* definiu `@person` e o nome dessa pessoa é Henry, um formulário contendo:
```

```erb
<%= text_field(:person, :name) %>
```

produz uma saída semelhante à 

```erb
<input id="person_name" name="person[name]" type="text" value="Henry" />
```

Após o envio do formulário, o valor inserido pelo usuário fica armazenado em `params[:person][:name]`.

WARNING: Você deve passar o nome da instância da variável, i.e. `:person` ou `"person"`, não a instância atual do objeto *model*.

O Rails fornece *helpers* para exibição de validação de erros associados ao objeto *model*. Esses são abordados em detalhes no guia de [Validação do *Active Record*](active_record_validations.html#displaying-validation-errors-in-views).

> ### Vinculando um formulário a um objeto

Reviewer 2: ```suggestion
### Vinculando um Formulário a um Objeto
```

Acho que podemos manter as iniciais maiúsculas só pra manter o padrão, o que acha?

<% end %>
```

Existem algumas coisas a serem observadas aqui:

* `@article` é o objeto real que está sendo editado.
* Existe um único *hash* de opções. Opções HTML (exceto `id` e `class`) são passadas no *hash* `:html`. Além disso, você pode fornecer uma opção `:namespace` no seu formulário para garantir a exclusividade dos atributos de *id* nos elementos do formulário. O atributo de escopo será prefixado com sublinhado no id HTML gerado.
* O método `form_with` produz um objeto **construtor de formulário** (a variável `f`).

> * Se você deseja direcionar sua requisição de formulário para um URL específico, use `form_with url: my_nifty_url_path`. Para ver as opções mais detalhadas sobre o que `form_with` aceita, [verifique a documentação da API](https://api.rubyonrails.org/classes/ActionView/Helpers/FormHelper.html#method-i-form_with).

Reviewer: ```suggestion
* Se você deseja direcionar sua requisição de formulário para uma URL específica, use `form_with url: my_nifty_url_path`. Para ver as opções mais detalhadas sobre o que `form_with` aceita, [verifique a documentação da API](https://api.rubyonrails.org/classes/ActionView/Helpers/FormHelper.html#method-i-form_with).
```

</form>
```

> O objeto passado como `:model` em `form_with` controla a chave usada em `params` para acessar os valores do fomulário. Aqui está o nome `article` e, portanto, todas as entradas tem nomes do formulário `article[attribute_name]`. Assim, na ação `create` haverá um *hash* `params[:article]` com as chaves `:title` e `:body`. Você pode ler mais sobre a importância dos nomes de entrada n capítulo [Compreendendo as convenções de nomemclatura de parametros](#understanding-parameter-naming-conventions) deste guia.

Reviewer: ```suggestion
O objeto passado como `:model` em `form_with` controla a chave usada em `params` para acessar os valores do fomulário. Aqui está o nome `article` e, portanto, todas as entradas tem nomes do formulário `article[attribute_name]`. Assim, na ação `create` haverá um *hash* `params[:article]` com as chaves `:title` e `:body`. Você pode ler mais sobre a importância dos nomes de entrada no capítulo [Entendendo Convenções de Nomeação de Parâmetros](#entendendo-convencoes-de-nomeacao-de-parametros) deste guia.
```

Aqui é só pra manter a coerência com o que já está traduzido, apesar de eu achar "nomenclatura" melhor que "nomeação".

https://guiarails.com.br/form_helpers.html#entendendo-convencoes-de-nomeacao-de-parametros

TIP: Convencionalmente, suas entradas espelharão os atributos do *model*. No entanto, eles não precisam! Se houver outras informações de que você precisa, você pode incluí-las em seu formulário da mesma forma que os atributos e acessá-las via `params[:article][:my_nifty_non_attribute_input]`.

Os métodos *helper* chamados no construtor de formulário (*form builder*) são identicos ao objeto *model helper*, exceto que não é necessário especificar qual objeto está sendo editado, pois isso já é gerenciado pelo construtor de formulário.

> Você pode criar uma vinculação semelhante sem realmente criar uma tag `<form>` com o *helper* `fields_for`. Isso é útil para editar objetos *model* adicionais com o mesmo formulário. Por exemplo, se você tem um *model* `Person` vínculado à um *model* `ContactDetail`, você pode criar um formulário para criar os dois, assim:

Reviewer 2: ```suggestion
Você pode criar uma vinculação semelhante sem realmente criar uma tag `<form>` com o *helper* `fields_for`. Isso é útil para editar objetos *model* adicionais com o mesmo formulário. Por exemplo, se você tem um *model* `Person` vinculado à um *model* `ContactDetail`, você pode criar um formulário para criar os dois, assim:
```

Aqui só tirei um typo, tinha um acento em vinculado. Pode conferir?

</form>
```

O objeto produzido por `fields_for` é um construtor de formulário igual ao produzido por `form_with`.

### Confiando na Identificação de Registro

> O *model* Article está disponível para os usuários da aplicação, portanto - seguindo as melhores praticas para desenvolvimento com Rails - você deve declará-lo **um recurso**:

Reviewer 2: ```suggestion
O *model* Article está disponível para os usuários da aplicação, portanto - seguindo as melhores praticas para desenvolvimento com Rails - você deve declará-lo **um recurso (*resource*)**:
```

Acho que aqui podemos incluir também o termo em ingles pra manter a leitura, o que acha?

```ruby
resources :articles
```

> TIP: Declarar um recurso tem vários efeitos colaterais. Consulte [o guia Rails *Routing from the Outside In*](routing.html#resource-routing-the-rails-default) para obter mais informações como configurar e usar recursos.

Reviewer: ```suggestion
TIP: Declarar um recurso tem vários efeitos colaterais. Consulte [o guia Rotas do Rails de Fora pra Dentro](routing.html#roteando-resources-recursos-o-padrao-do-rails) para obter mais informações como configurar e usar recursos.
```

Aqui também já temos a seção traduzida
https://guiarails.com.br/routing.html#roteando-resources-recursos-o-padrao-do-rails

form_with(model: @article)
```

> Observe como a `form_with` chamada de estilo abreviado é convenientemente a mesma, independentemente de o registro ser novo ou existente. A identificação de registros é inteligente o suficiente para descobrir se o registro é novo, chamando `record.persisted?`. Ele também seleciona o caminho correto para enviar e o nome com base na classe do objeto.

Reviewer: ```suggestion
Observe como a chamada de estilo abreviado `form_with` é convenientemente a mesma, independentemente de o registro ser novo ou existente. A identificação de registros é inteligente o suficiente para descobrir se o registro é novo, chamando `record.persisted?`. Ele também seleciona o caminho correto para enviar e o nome com base na classe do objeto.
```

Aqui a sugestão é apenas para inverter a ordem das palavras. O que acha? :smile:

> WARNING: Quando você está usando herança de tabela única  (STI ou *single-table inheritance*) em seus objetos *model*, você não pode confiar na identificação de registro em uma subclasse se apenas sua classe pai for declarada um recurso. Você terá que especificar `:url`, e `:scope` (o nome do *model*) explicitamente.

Reviewer 2: ```suggestion
WARNING: Quando você está usando herança de tabela única (STI ou *single-table inheritance*) em seus objetos *model*, você não pode confiar na identificação de registro em uma subclasse se apenas sua classe pai for declarada um recurso. Você terá que especificar `:url`, e `:scope` (o nome do *model*) explicitamente.
```

Aqui tinha um pequeno espaço a mais nada demais 😄

#### Trabalhando com *Namespaces*

Se você criou rotas com *namespaces*, `form_with` tem uma abreviatura bacana para isso também. Se o seu aplicativo tiver um *namespace* de administrador,

```ruby
form_with model: [:admin, @article]
```

irá criar um formulário que submete ao `ArticlesController` dentro do *namespace* do administrador (submetendo a `admin_article_path(@article)` no caso de uma atualização). Se você tiver vários níveis de *namespacing*, a sintaxe será semelhante:

```ruby
form_with model: [:admin, :management, @article]
```

> Para mais informações sobre o sistema de roteamento Rails e as convenções associadas, consulte o [guia Rails Routing from the Outside In](routing.html).

Reviewer: ```suggestion
Para mais informações sobre o sistema de roteamento Rails e as convenções associadas, consulte o [guia Rotas do Rails de Fora pra Dentro](routing.html).
```

</form>
```

Ao analisar os dados submetidos pelo *POST*, o Rails levará em consideração o parâmetro especial `_method` e agirá como se o método HTTP fosse aquele especificado dentro dele (*"PATCH"* nesse exemplo).

> IMPORTANT: Todos os formulários usando `form_with` implementam `remote: true` por padrão. Esses formulários enviarão dados usando requisições XHR (*Ajax*). Para desabilitar isso, inclua `local: true`. Para se aprofundar, veja o guia [Trabalhando com JavaScript no Rails](working_with_javascript_in_rails.html#remote-elements).

Reviewer: ```suggestion
IMPORTANT: Todos os formulários usando `form_with` implementam `remote: true` por padrão. Esses formulários enviarão dados usando requisições XHR (*Ajax*). Para desabilitar isso, inclua `local: true`. Para se aprofundar, veja o guia [Trabalhando com JavaScript no Rails](working_with_javascript_in_rails.html#elementos-remotos).
```

https://guiarails.com.br/working_with_javascript_in_rails.html#elementos-remotos
